### PR TITLE
Rustdoc gui tests

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This is a reopening of #70533.

For this first version, there will be no screenshot comparison. Also, a big change compared to the previous version: the tests are now hosted in the rust repository directly. Since there is no image, it's pretty lightweight to say the least.

So now, only remains the nodejs script to run the tests and the tests themselves. Just one thing is missing: where should I put the documentation for these tests? I'm not sure where would be the best place for that. The doc will contain important information like the documentation of the framework used and how to install it (`npm install browser-ui-test`, but still needs to be put somewhere so no one is lost).

We'd also need to install the package when running the CI too. For now, it runs as long as we have nodejs installed, but I think we don't it to run in all nodejs targets?

cc @jyn514 

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Also: weird CI failure...

**Comment by jyn514** [comment]
> where should I put the documentation for these tests? 

I think either a README in the directory they're in, or in the dev-guide would be a good place. Possibly one could link to the other.

File: src/tools/rustdoc-gui/tester.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
+const {Options, runTest} = require('browser-ui-test');
```
Comment (jyn514) [inline src/tools/rustdoc-gui/tester.js:1]: Where does this package come from? Does it need to be installed by the user or by bootstrap?

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:1]: I said it in the PR message that it was part of the information I wanted to put inside the documentation. I also gave the command to install it (which is `npm install browser-ui-test`). ;)

Comment (jyn514) [reply src/tools/rustdoc-gui/tester.js:1]: Oops, I should read :laughing: thanks!

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:1]: No, I think it'd be nice to add a comment in the code actually. I'll do that.

File: src/bootstrap/test.rs [hunk 2]
```diff
@@ -688,6 +688,57 @@ impl Step for RustdocJSNotStd {
...
+
+    fn run(self, builder: &Builder<'_>) {
+        if let Some(nodejs) = &builder.config.nodejs {
+            builder.ensure(compile::Std { compiler: self.compiler, target: self.target });
+            let out_dir = builder.test_out(self.target).join("rustdoc-gui");
+            let mut command = builder.rustdoc_cmd(self.compiler);
+            command.arg("src/test/rustdoc-gui/lib.rs").arg("-o").arg(&out_dir);
+            builder.run(&mut command);
+
+            for file in fs::read_dir("src/test/rustdoc-gui").unwrap() {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:719]: This needs something like `builder.src.join("src/test/rustdoc-gui")` -- my guess is probably the same should be true of the src/test/rustdoc-gui/lib.rs before, not sure why that one is working.

**Comment by Mark-Simulacrum** [comment]
I am still a little uncertain about the approach here of needing a JS out-of-tree dependency expected to be installed through nodejs/npm; I definitely think this cannot land without some amount of published documentation on how to update these tests. I am a bit worried that it's going to be a significant increase in the overhead for contributors to rustdoc (needing a JS toolchain installed to test things locally), but I could be wrong about that. Ultimately that's a T-rustdoc decision, I suppose.

The comment I left should hopefully fix CI.

**Comment by GuillaumeGomez** [comment]
We can clone the framework too. I don't have much preference for a solution, just thought that it might be simpler to use `npm` to do so. In the previous version, it's true that we cloned the repository inside `bootstrap` script and then run it. So whatever way you prefer.

**Comment by Mark-Simulacrum** [comment]
Even cloning the framework presumably doesn't decouple us from needing the JS toolchain though? I think that's my primary concern. If it's just nodejs that's not too bad, but I'm a bit worried about running npm install mid-build.

**Comment by GuillaumeGomez** [comment]
Then let's go for a submodule. Like that at we'll know what's going on and no need for npm.

**Comment by GuillaumeGomez** [comment]
To make things a bit simpler to track down, I made the `browser-UI-test` into a submodule instead of relying completely on `npm`. However, we still need to install `puppeteer` and `pngjs` with `npm`. Another concern that @Mark-Simulacrum had was that the test suite would be run too often, so to prevent this to happen, I added the following checks:

 * You need both `npm` and `nodejs` to be installed
 * You need the `browser-UI-test` dependencies to be installed (through `npm`) as well

In case only the second condition isn't respected, a warning is displayed.

The last commit is the CI update. I took the changes mostly from #70533 but we'll still need to have @Mark-Simulacrum to confirm them. :)

So once everything else has been validated, I'll add the documentation.

**Comment by Mark-Simulacrum** [comment]
If we're going to require npm anyway I see no reason for the submodule I think? We probably want to put the lockfile in git though...

**Comment by GuillaumeGomez** [comment]
That's the thing: either we put a lockfile in rust repository, or we keep the submodule and use its own lockfile. One way or another, it's fine by me. Maybe you have an opinion @jyn514 ?

**Comment by Mark-Simulacrum** [comment]
I have several requirements:

 * we should use published (to npm or crates.io etc) deps or rust-lang repositories in submodules, to help ensure they don't go away on us; regardless a lockfile of some kind should exist.
 * it should be easy to ensure that the test here is run by CI (e.g. minimal checking for just node is fine); if we start checking for the presence of multiple binaries I'd rather just gate on a new config.toml option

It's up to the rustdoc team whether you think this requirement is a reasonable ask of contributors - it seems like a high bar to me, but maybe editing these tests is rare and not too worrisome.

I will take a look at this PR's implementation soon.

**Comment by GuillaumeGomez** [comment]
These tests are mostly for me. To be more precise: I spend most of my time fixing GUI issues and I'd that to slow down quite a lot. So unless someone is changing the front-end, they don't need to run this test suite, and even if they do, I'm pretty sure in most cases there won't be an issue (but if there is one, the CI will tell us, finally!).

**Comment by jyn514** [comment]
@GuillaumeGomez that's not how testing works, though. CI can fail on anyone's changes, not just yours. The scenario is someone *does* make changes and either needs to add a new test or an existing test fails.

That said, I think if we print "run `npm install x`" if the package isn't installed, that's good enough - I'd expect anyone modifying the frontend to at least be familiar with npm even if it's not currently installed.

**Comment by GuillaumeGomez** [comment]
I maybe badly tried to explain my point of view. The idea was "if you broke the GUI tests, you very likely know how npm works". :smile:

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum I might need help for the CI scripts update (not too sure what to change here without breaking everything...). So whenever you have time, it'd be super appreciated! :)

**Comment by Mark-Simulacrum** [comment]
I think this is still adding a submodule which seems like the wrong approach. I would expect us to use browser-ui-test just like we expect e.g. a C compiler to be installed to run this test. I presume that it can be readily installed by contributors via npm or so?

**Comment by GuillaumeGomez** [comment]
Hmmm... Well, it doesn't hurt to have `package.json` and `package-lock.json` files instead of a submodule I guess. I'll make the update soon. But still need your help for the CI files though. ;)

**Comment by Mark-Simulacrum** [comment]
I don't think we would expect to have lockfiles in-tree for "system tooling" - if the UI test framework you're proposing we use is no different to a C compiler, it seems odd to require a very specific version.

CI should be a matter of installing the necessary dependency (similar to what you're doing now, I expect).

**Comment by GuillaumeGomez** [comment]
Maybe? I was looking at the UI framework repository and saw both files. If the lock file is unnecessary I will not add it.

**Comment by jyn514** [comment]
> I think this is still adding a submodule which seems like the wrong approach. I would expect us to use browser-ui-test just like we expect e.g. a C compiler to be installed to run this test. I presume that it can be readily installed by contributors via npm or so?

I don't think a C compiler is the right analogy - I think libreadline would be a better comparison, where you need *some* version, and it's not feasible to install it yourself, but the exact version doesn't matter.

That said, I don't think `npm install -g browser-ui-test` is much easier or harder than `npm install` with a package.json, as long as it's documented.

**Comment by Mark-Simulacrum** [comment]
Sure, libreadline and a C compiler feel pretty similar to me in this case. I think I would prefer a global install over package.json in the main repo (though we can check in the lockfile as part of the docker directory, that seems appropriate).

**Comment by GuillaumeGomez** [comment]
I don't think this is the right approach in here for the simple reason that `browser-ui-test` will very likely evolve depending on our needs. Now imagine we add a new command/feature and that someone with an older version tries to run it: it'll simply crash because of the unknown command/feature and finding out what's the issue could be quite tricky. So having a submodule or package.json seems simpler to me (well, the submodule mostly because when out-of-date, you see it right away).

**Comment by Mark-Simulacrum** [comment]
We can definitely detect the version and error out (like we do for LLVM, gdb/lldb, probably other tools). But I am unconvinced that a tool that is -- as far as I can tell -- developed *not* for rustdoc, but rather intended to be more general, should be versioned along with its dependencies in this repository. I am also concerned with adding JS-specific lockfiles etc to the root of the repository, because they're likely not necessary for almost all developers right now and I would not want to create additional confusion as to what you need to contribute to this repository. If the framework *is* specific to rustdoc, then in some sense I would probably want to know why we can't use something more off the shelf  - if only because I am personally interested in similar testing elsewhere, and if there's really nothing out there that we can reuse (e.g., like we do LLVM's FileCheck for codegen tests), that's useful to know.

**Comment by GuillaumeGomez** [comment]
I don't like the idea of adding JS config/lock files either, which is why I first suggested adding submodule. I just wanted to answer to your last reply because it didn't seem much better. Anyway, we can go with that as well, I'm fine with it!

**Comment by JohnCSimon** [comment]
@JohnCSimon label: +S-waiting-on-review -S-waiting-on-author

**Comment by JohnCSimon** [comment]
`@JohnCSimon label: +S-waiting-on-review -S-waiting-on-author`

**Comment by JohnCSimon** [comment]
@Rustbot label: +S-waiting-on-review -S-waiting-on-author

**Comment by GuillaumeGomez** [comment]
I removed the submodule, the install will only go through `npm` now.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum I now need your help for the CI. :)

**Comment by jyn514** [comment]
@GuillaumeGomez you need to run `./x.py run src/tools/expand-yaml-anchors`

**Comment by Mark-Simulacrum** [comment]
I've not taken a look at this yet but will try to do so in the coming week. If you could provide a summary of what you believe to be open questions and particular areas for me to pay attention to that would be quite helpful.

I also filed #82013 to try and clarify the CI error to help the user out.

**Comment by GuillaumeGomez** [comment]
@jyn514 Running that command broke the CI config haha.

@Mark-Simulacrum: I need help for the CI files in order to ensure that the GUI tests are run. I think that's pretty much it. :)

**Comment by GuillaumeGomez** [comment]
It's finally working! \o/

**Review by Mark-Simulacrum — COMMENTED** [review]
CI change (to the try builder) also needs to be reverted before we merge.

File: src/ci/docker/host-x86_64/x86_64-gnu-aux/Dockerfile [hunk 3]
```diff
@@ -17,10 +17,26 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
+  fonts-kacst \
+  fonts-freefont-ttf \
+  libxss1 \
+  libxtst6
+
+RUN curl -sL https://nodejs.org/dist/v14.4.0/node-v14.4.0-linux-x64.tar.xz | tar -xJ
+ENV PATH="/node-v14.4.0-linux-x64/bin:${PATH}"
+
+# Install required dependencies from browser-UI-test framework
+RUN npm install browser-ui-test -g --unsafe-perm=true
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-aux/Dockerfile:36]: I'll add a complete comment about it.

**Comment by GuillaumeGomez** [comment]
> CI change (to the try builder) also needs to be reverted before we merge.

Which changes are you referring to btw?

**Comment by Mark-Simulacrum** [comment]
.github/workflows/ci.yml is modified.

**Comment by GuillaumeGomez** [comment]
I added an explanation (and a link to the issue on the puppeteer repository) and reverted the changes in the `.github/workflows/ci.yml` file.

**Comment by Mark-Simulacrum** [comment]
src/ci/github-actions/ci.yml was not reverted. Changes otherwise look good, but please squash commits; r=me with that done.

@bors rollup=never

**Comment by GuillaumeGomez** [comment]
Commits squashed, `src/ci/github-actions/ci.yml` changes were removed as well. Thanks a lot for your help!

@bors: r=Mark-Simulacrum

**Comment by GuillaumeGomez** [comment]
There is no failure apparently? Well, let's rety...

@bors: retry

**Comment by ehuss** [comment]
@bors retry
macos: system is shutting down

File: .gitignore [hunk 4]
```diff
@@ -50,4 +50,6 @@ Session.vim
 .cargo
 !/src/test/run-make/thumb-none-qemu/example/.cargo
 no_llvm_build
+**node_modules
+**package-lock.json
```
Comment (jethrogb) [inline .gitignore:54]: I don't think this is valid syntax. At least I can't find it in `man gitignore` and ripgrep complains about it:
```
rust/.gitignore: line 55: error parsing glob '**node_modules': invalid use of **; must be one path component
rust/.gitignore: line 56: error parsing glob '**package-lock.json': invalid use of **; must be one path component
```

Comment (GuillaumeGomez) [reply .gitignore:54]: `**` is to look for all subfolders, whatever the level.

Comment (jethrogb) [reply .gitignore:54]: I believe that would be `**/node_modules`

Comment (GuillaumeGomez) [reply .gitignore:54]: It works for me as is but don't hesitate to send an update.

**Pull request merged**